### PR TITLE
LTS-13

### PR DESCRIPTION
**Pull request by domenkozar**
# TODO

- [x] servant-client-streaming client integration

### CODE REVIEW

File: cachix/src/Cachix/Client/Commands.hs [hunk 1]
```diff
@@ -238,58 +235,57 @@ pushStorePath env name storePath = retryPath $ \retrystatus -> do
...
-        -- Upload narinfo with signature
-        NoContent <- escalate <=< (`runClientM` clientenv env) $ Api.createNarinfo
-          (cachixBCClient name)
-          (Api.NarInfoC storeHash)
-          nic
-        return ()
+        void $ (`withClientM` newClientEnv)
+            (Api.createNar (cachixBCClient name) stream')
+            $ \resp -> case resp of
+              Left _ -> escalate resp
```
Comment (roberth) [inline cachix/src/Cachix/Client/Commands.hs]: I suggest using escalate instead of pattern matching here

**Comment by sib-null** [comment]
Can this be merged yet?

**Comment by domenkozar** [comment]
Blocked on https://github.com/input-output-hk/stack2nix/pull/155

**Comment by domenkozar** [comment]
Blocked on https://gitlab.com/twittner/bytestring-conversion/issues/2 for darwin

**Comment by domenkozar** [comment]
Should be all fixed for LTS-13.4, so expect a release early next week.

File: cachix-api/src/Cachix/Api.hs [hunk 2]
```diff
@@ -55,10 +58,10 @@ data BinaryCacheAPI route = BinaryCacheAPI
   , nar :: route :-
       "nar" :>
       Capture "nar" NarC :>
-      StreamResponseGet '[XNixNar, JSON]
+      StreamGet NoFraming OctetStream (ConduitT () ByteString IO ())
   , createNar :: route :-
       "nar" :>
-      StreamBodyMonad '[XNixNar, JSON] AWS :>
+      StreamBody NoFraming OctetStream (ConduitT () ByteString AWS ()) :> -- XNixNar
```
Comment (roberth) [inline cachix-api/src/Cachix/Api.hs]: Is the cachix-api conduit dependency avoidable?

File: cachix-api/src/Cachix/Types/SwaggerOrphans.hs [hunk 3]
```diff
@@ -1,31 +1,23 @@
...
-#endif
 
--- https://github.com/plow-technologies/servant-streaming/blob/master/servant-streaming-docs/src/Servant/Streaming/Docs/Internal.hs
--- TODO: these should define the body/response content
-instance (HasSwagger api) => HasSwagger (StreamBodyMonad contentTypes m :> api) where
-  toSwagger _ = toSwagger (Proxy :: Proxy api)
+-- TODO: upstream to servant-conduit
+instance ToSchema i => ToSchema (ConduitT i o m r) where
+  declareNamedSchema _ = declareNamedSchema (Proxy :: Proxy i)
+  -- TODO: Proxy o
```
Comment (roberth) [inline cachix-api/src/Cachix/Types/SwaggerOrphans.hs]: Looks like a nasty mismatch if you want to have a schema for the output. Does upstream have an issue about this?

File: cachix/src/Cachix/Client/Commands.hs [hunk 4]
```diff
@@ -238,58 +235,55 @@ pushStorePath env name storePath = retryPath $ \retrystatus -> do
...
-
-        -- Upload narinfo with signature
-        NoContent <- escalate <=< (`runClientM` clientenv env) $ Api.createNarinfo
-          (cachixBCClient name)
-          (Api.NarInfoC storeHash)
-          nic
-        return ()
+        void $ (`withClientM` newClientEnv)
+            (Api.createNar (cachixBCClient name) stream')
+            $ escalate >=> \NoContent -> do
```
Comment (roberth) [inline cachix/src/Cachix/Client/Commands.hs]: I don't like this combination of right-to-left and left-to-right operators.
What was wrong with the `NoContent <-` binding?

File: cachix-api/src/Cachix/Api.hs (hunk 2 above)
Comment (domenkozar) [reply cachix-api/src/Cachix/Api.hs]: Not that I know of. Same for AWS.

File: cachix/src/Cachix/Client/Commands.hs (hunk 4 above)
Comment (domenkozar) [reply cachix/src/Cachix/Client/Commands.hs]: It's a different API now, it is CPS so it's either this or the previous way you wanted to get rid of double `resp`

Comment (roberth) [reply cachix/src/Cachix/Client/Commands.hs]: :+1:

**Comment by domenkozar** [comment]
Ready from my side.

**Pull request merged**